pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/httproute_reconciler.go

```diff
@@ -704,45 +725,65 @@ func (r *RawHTTPRouteReconciler) reconcileTopLevelHTTPRoute(ctx context.Context,
-		return err
-	}
-	// reconcile httpRoute
+
 	existingHttpRoute := &gatewayapiv1.HTTPRoute{}
-	err = r.client.Get(ctx, types.NamespacedName{
+	getExistingErr := r.client.Get(ctx, types.NamespacedName{
 		Namespace: isvc.Namespace,
 		Name:      isvc.Name,
 	}, existingHttpRoute)
-	if err != nil {
-		if apierr.IsNotFound(err) {
-			log.Info("Creating top level HttpRoute resource", "name", isvc.Name)
-			if err := r.client.Create(ctx, desired); err != nil {
-				log.Error(err, "Failed to create top level HttpRoute", "name", desired.Name)
-				return err
-			}
-		} else {
-			return err
+
+	if !isvc.GetForceStopRuntime() {
+		if desired == nil {
+			return nil
 		}
-	} else {
-		// Set ResourceVersion which is required for update operation.
-		desired.ResourceVersion = existingHttpRoute.ResourceVersion
-		// Do a dry-run update to avoid diffs generated by default values.
-		// This will populate our local httpRoute with any default values that are present on the remote version.
-		if err := r.client.Update(ctx, desired, client.DryRunAll); err != nil {
-			log.Error(err, "Failed to perform dry-run update on top level httpRoute", "name", desired.Name)
+
+		if err := controllerutil.SetControllerReference(isvc, desired, r.scheme); err != nil {
+			log.Error(err, "Failed to set controller reference for top level HttpRoute", "name", isvc.Name)
 			return err
 		}
-		if !semanticHttpRouteEquals(desired, existingHttpRoute) {
-			if err = r.client.Update(ctx, desired); err != nil {
-				log.Error(err, "Failed to update toplevel HttpRoute", "name", isvc.Name)
+
+		// reconcile httpRoute
+		if getExistingErr != nil {
+			if apierr.IsNotFound(getExistingErr) {
+				log.Info("Creating top level HttpRoute resource", "name", isvc.Name)
+				if err := r.client.Create(ctx, desired); err != nil {
+					log.Error(err, "Failed to create top level HttpRoute", "name", desired.Name)
+					return err
+				}
+			} else {
+				return getExistingErr
+			}
+		} else {
+			// Set ResourceVersion which is required for update operation.
+			desired.ResourceVersion = existingHttpRoute.ResourceVersion
+			// Do a dry-run update to avoid diffs generated by default values.
+			// This will populate our local httpRoute with any default values that are present on the remote version.
+			if err := r.client.Update(ctx, desired, client.DryRunAll); err != nil {
+				log.Error(err, "Failed to perform dry-run update on top level httpRoute", "name", desired.Name)
 				return err
 			}
+			if !semanticHttpRouteEquals(desired, existingHttpRoute) {
+				if err = r.client.Update(ctx, desired); err != nil {
+					log.Error(err, "Failed to update toplevel HttpRoute", "name", isvc.Name)
+					return err
+				}
+			}
+		}
+	} else {
+		// Delete the http route
+		if getExistingErr == nil {
+			// Make sure that we only delete http route owned by the isvc
+			if existingHttpRoute.OwnerReferences[0].UID == isvc.UID && existingHttpRoute.GetDeletionTimestamp() == nil {
```

Comment:
is it guaranteed in every case that the first owner reference is "ours"? if it is, to make it more robust to future changes, should we loop through each owner ref?
don't we need to check the length before trying to access the first element?

Reply:
@israel-hdez Can you weigh in? I think I asked a similar question about needing to loop through the references when you suggested adding checking for the ownership before deleting, but I don't remember the logic that allowed `OwnerReferences[0]` to be okay